File: 5. Longest Palindromic Substring/solution2.js
/**
 * @param {string} s
 * @return {string}
 */

// This function returns the longest palindromic substring in s
// A palindrome is a string that is the same when read forward or backward
// For example, "racecar" and "abba" are palindromes
var longestPalindrome = function(s) {
    // Initialize a variable to store the longest palindrome found so far
    var max = '';
    // Loop through each character in s
    for (var i = 0; i < s.length; i++) {
      // This loop is to take into account 
      // different types of palindromes like 'aba' and 'abba'
      // We will try to expand around the center of each character
      // and check if the substring is a palindrome
      for (var j = 0; j < 2; j++) {
        // Set the left and right pointers to the center of the current character
        var left = i;
        var right = i + j;
        // While the left and right pointers are within the bounds of s
        // and the characters at those positions are equal
        while (s[left] && s[left] === s[right]) {
          // Move the left pointer one step to the left
          left--;
          // Move the right pointer one step to the right
          right++;
        }
        // At this point, we have either reached the end of s
        // or found a mismatching character
        // The palindrome is the substring from left + 1 to right - 1
        // Note that we need to adjust the indices because of the previous steps
        // For example, if s = "sabbad", then after finding "abba"
        // we have left = 0 and right = 5, but the actual indices of "abba" are 1 and 4
        // So we need to add 1 to left and subtract 1 from right
        // The length of the palindrome is then right - left - 1
        if ((right - left - 1) > max.length) {
          // If the length of the palindrome is greater than max.length
          // then update max to be that palindrome
          max = s.substring(left + 1, right);
        }
          
      }
    }
    // Return the longest palindrome found in s
    return max;
  };
  
  // The time complexity of this solution is O(N^2), where N is the length of s
  // This is because we loop through each character in s, and for each character
  // we expand around the center up to O(N) times in the worst case
  // The space complexity of this solution is O(1), because we only use a constant amount of extra space to store max
  

File: 76. Minimum Window Substring/solution2.cpp
class Solution {
public:
    string minWindow(string s, string t) {
        // create a vector of size 128 to store the frequency of each character in t
        vector<int> map(128,0);

        // loop through each character in t and increment its frequency in map
        for (char c : t) {
            map[c]++;
        }

        // initialize counter to the size of t, which is the number of characters we need to find in s
        int counter = t.size(), begin = 0, end = 0, d = INT_MAX, head = 0;

        // loop through each character in s using end as the pointer
        while (end < s.size()){

            // if the current character is in t, decrement its frequency in map and decrement counter
            // this means we have found one more character in the current window
            if (map[s[end++]]-- > 0) {
                counter--;
            }

            // while counter is zero, which means we have found all the characters in t in the current window
            while (counter == 0) {

                // if the current window is smaller than the previous minimum window, update head and d
                if (end - begin < d) {
                    head = begin;
                    d = end - head;
                }

                // if the character at begin is in t, increment its frequency in map and increment counter
                // this means we are moving the begin pointer to shrink the window and look for a smaller one
                if (map[s[begin++]]++ == 0) {
                    counter++;
                }
            }  
        }

        // if d is INT_MAX, it means we did not find any window that contains all the characters in t, so return an empty string
        // otherwise, return the substring of s from head to head + d, which is the minimum window
        return d == INT_MAX ? "" : s.substr(head, d);
    }
};

/* 
Complexity Analysis:
    The time complexity of this code is O(n), where n is the length of s, because we only scan through s once using the two pointers. 
    The space complexity of this code is O(1), because we only use a constant size vector map and a few variables.
 */